XRPLib/imu.py

```diff
@@ -356,22 +399,34 @@ def calibrate(self, calibration_time:float=1, vertical_axis:int= 2, update_time:
         :type update_time: int
         """
         self.update_timer.deinit()
-        start_time = time.ticks_ms()
         self.acc_offsets = [0,0,0]
         self.gyro_offsets = [0,0,0]
         avg_vals = [[0,0,0],[0,0,0]]
         num_vals = 0
+        # Wait a bit for sensor to start measuring (data registers may default to something nonsensical)
+        time.sleep(.1)
+        start_time = time.ticks_ms()
         while time.ticks_ms() < start_time + calibration_time*1000:
             cur_vals = self._get_acc_gyro_rates()
             # Accelerometer averages
-            avg_vals[0][0] = (avg_vals[0][0]*num_vals+cur_vals[0][0])/(num_vals+1)
-            avg_vals[0][1] = (avg_vals[0][1]*num_vals+cur_vals[0][1])/(num_vals+1)
-            avg_vals[0][2] = (avg_vals[0][2]*num_vals+cur_vals[0][2])/(num_vals+1)
+            avg_vals[0][0] += cur_vals[0][0]
+            avg_vals[0][1] += cur_vals[0][1]
+            avg_vals[0][2] += cur_vals[0][2]
             # Gyroscope averages
-            avg_vals[1][0] = (avg_vals[1][0]*num_vals+cur_vals[1][0])/(num_vals+1)
-            avg_vals[1][1] = (avg_vals[1][1]*num_vals+cur_vals[1][1])/(num_vals+1)
-            avg_vals[1][2] = (avg_vals[1][2]*num_vals+cur_vals[1][2])/(num_vals+1)
-            time.sleep(0.05)
+            avg_vals[1][0] += cur_vals[1][0]
+            avg_vals[1][1] += cur_vals[1][1]
+            avg_vals[1][2] += cur_vals[1][2]
+            # Increment counter and wait for next loop
+            num_vals += 1
+            time.sleep(update_time / 1000)
+
+        # Compute averages
+        avg_vals[0][0] /= num_vals
+        avg_vals[0][1] /= num_vals
+        avg_vals[0][2] /= num_vals
+        avg_vals[1][0] /= num_vals
+        avg_vals[1][1] /= num_vals
+        avg_vals[1][2] /= num_vals
 
         avg_vals[0][vertical_axis] -= 1000 #in mg
```

Comment:
I would actually refrain from implementing a gravitational offset in the library, at least by default. Gravity can be really useful for determining the pitch and roll of the robot, freefall detection, etc. Plus it's really easy for the user to implement this offset by themselves if they need it.

Reply:
The gravitational offset here is because we initially assume all values are 0 by default when taking the averages that we use for the offets. We then have to go back and correct one of the axis so that the gravity vector is actually visible in the readings, otherwise all we'll get after calibration is zeros

Reply:
Oh I see, I misunderstood what the code was doing, my bad!

Hmm, thinking about it, should we really include the accelerometer in the offset calibration? This assumes the board is perfectly flat with respect to gravity, but that definitely won't be the case for all users. How far off are the measurements with no offset? Is an offset actually needed? Can we perhaps separate the gryo and accelerometer calibration, then leave it up to the user to run the accelerometer calibration? Could create a separate `imu.calibrate_accel()` and allow the user to pass in the orientation of their board.

Reply:
There's definitely a non-negligible offset from the accelerometer readings. It doesn't need to be as carefully calibrated as the gyro bc we're not really integrating it, so drift doesn't really matter, but we should definitely be calibrating it at least a little before letting the students use it.

Separating the gyro and accelerometer calibration is an interesting idea, but passing in the orientation (or using the gyro) is still based in them having calibrated the gyro on a flat surface. I honestly think it's probably fine to say that most people will be calibrating their bots on a flat (or almost flat) surface, and that the slight deviation from that will have a relatively small impact.

The current implementation allows for people to calibrate their board using either the X, Y, or Z axis as the one starting with gravity positive, but the more I think about it the less use cases I can think for this option. It may just be easier (and less confusing to someone looking at the documentation) if we were to just remove that option

Reply:
Fair points! I had a couple other thoughts come to mine:

Do we need to calibrate the IMU every single time, or can we do it once and store the offsets in a file or something? That'd help reduce the down time before user code executes, and make the calibration less prone to the user moving the robot during the calibration.

On a related note, what if the user doesn't know about the calibration routine, and they move the robot during it? I suppose we could detect it by evaluating the accel/gyro values, but then what? I'm wondering if it would be better to have the user call the calibrate function themselves, or something that bring the users attention to the fact that they shouldn't move the robot during that time.

I also think it may be worth revisiting this topic if/when an AHRS algorithm is implemented. In that case, the accelerometer would be used to determine the pitch and roll angles, so might affect how we decide to calibrate it (eg. if the user starts the robot on a surface tilted at 10 degrees, should the IMU say that's 10 degrees or zero degrees?)

Reply:
Storing calibration data is an interesting idea, definitely worth exploring it. 

Atm we're calibrating the gyro immediately for ease of use, since the default functionality is for the drivetrain to use it. We could change that, but I think it's useful for the drivetrain to just work when using it. What if we were to move the calibrate call from `get_default_imu` to `get_default_drivetrain`, so if a user wanted to use the IMU on its own they can control when exactly it calibrates? although they can just call calibrate on their own anyways. 

I'm not sure users knowledge of the calibration routine is a huge issue, since the bots are being coupled with a class and we can just makes sure the teachers pass that knowledge along. Would it be worthwhile to add a print statement or something that just says "calibrating robot, please don't move the robot during this time" and "Finished calibration, program running"?

As for that last note, I'm not sure which behavior is preferable. I've asked a few people what their personal thoughts are and overall I'm thinking that when you calibrate, all of your angles should be relative to that point. I'm going to be trying to implement the AHRS algorithm fairly soon, so we can revisit that discussion then.

Reply:
My main concern is that many (most?) people won't even know there's a calibration that occurs, and that the robot needs to be completely motionless during it. I could totally foresee a lot of people moving the robot around during that time for whatever reason (eg. carrying the robot while turning it on, or positioning it on a line just after clicking the run button in the IDE, or lightly bumping the robot, or...), then having their robot not drive straight or not turn to the correct angles. That will inevitably cause confusion, and will probably make some people think the robot is just bad, or theirs is broken. I can't count how many times I've seen hidden behavior like this trip up students.

I wouldn't require teachers relay that information (they've got enough to do already), plus there's a lot more use cases for the XRP than just classes. IMO it should be trivial for a user to discover that there is a calibration routine, and that the robot needs to be motionless during it. At a minimum, I think adding some prints statements before and after would be good. I think it'd also help to have it documented in multiple places (eg. the API documentation and the online curriculum). Or even better would be storing the calibration values in a file (could add a button in the IDE for this) so users don't have to worry about it all the time (assuming the IMU doesn't need to be calibrated frequently, need to test that).

Whether the angles should be relative or absolute with respect to gravity, this may actually depend on the AHRS algorithm. Most require the accelerometer ([list of algorithms here](https://ahrs.readthedocs.io/en/latest/#new-in-version-0-3)), and I would imagine those would bring the roll and pitch to zero if the accelerometer measures (x,y,z)=(0,0,1). I suppose you could add offsets for the roll and pitch if that's really desired. Idk if relative or absolute angles are best, but I'd personally lean towards absolute.

Also with regards to the accelerometer calibration, one other consideration is that the X and Y axis bounds should really be as close to +/- 1g as possible when the board is tilted sideways. After playing with my board, if the calibration is done on a non-level surface, the bounds can end up being wrong (eg. +1.1 and -0.9). So probably best to mention that the calibration should be done on a level surface, in addition to being motionless.

Reply:
Yea currently calibrating on a non-flat surface is gonna cause some issues atm, since we do the `vertical_axis -= 1000`, which assumes that gravity is aligned perfectly with one of the vectors. I'm still not happy with this solution, so i'm tempted to just say that it has to be flat, and to remove the option to change the vertical axis. I think we can still revisit this later, when the AHRS code is ready to be merged.

Regardless, I'm okay going through and testing how well saving calibration data to a file works myself, so I think we're all good to merge your changes in now

Reply:
Sounds good! Will leave it as is for now so we can merge this.